### ISSUE TITLE
Translation of Script Lab into other languages, e.g. German

### ISSUE THREAD

**Issue by maninweb** [OfficeDev/script-lab#398]
Hi there,

it would be great, if Script Lab could be also available in other languages. For German, I would like to offer my help, as I am native speaker for German. If you are interested, can you please tell me, how can I support you best and e.g. which files I can translate and how to submit them here? Thank you :-)

Best,
Mourad

**Comment by Zlatkovsky** [comment]
@maninweb , if you're willing to treat this as an experiment (and probably take a bit of time with this...), I'm happy to guide you through it.  And it would be a very useful thing both to the German-speaking folks, and also to us at Script Lab to both enable others to come after you with localization, and to discover any gaps in either our processes or our localizability.

Localization will come down to 2 things:  localizing the manifest (ribbon buttons), and the actual editor/runner UI.  The latter is more involved, but is also where 95% of the value is.  So let's focus on that one first, and assume that the ribbon stuff will be easy.

Your first step would be to follow [CONTRIBUTING.md](https://github.com/OfficeDev/script-lab/blob/master/CONTRIBUTING.md) to ensure that you can build and run Script Lab locally.  If you run into any issues, feel free to reach out (and also to submit a pull request with any changes to the doc, for places where we should have included more instructions).

Once you're able to run locally, as a 1st step, assume that you have _purely_ a German-speaking audience and replace the current English strings with German ones.  We'll add the redirection/switching at a later step.  For now, go ahead and localize anything in the two existing string file:

- client (editor):  https://github.com/OfficeDev/script-lab/blob/eeb4ba89a7a570cf1bb035fdd740ec7dcfb92822/src/server/core/strings.ts
- server (running snippet):  https://github.com/OfficeDev/script-lab/blob/eeb4ba89a7a570cf1bb035fdd740ec7dcfb92822/src/client/app/helpers/strings.ts

Having done that, step number 2 will be to see what strings are still left over that are English.  I would imagine that there are a bunch inside of the UI, both in:

- Angular templates within `src/client/app/components` and `src/client/app/containers`.  Pretty much anywhere that there is a `@Component`, there is a good chance there is some UI with English strings hiding therein.
- files in the `src/server/templates` folder
- `tutorial.html` and `run.html` in `src/client/views`

Depending on when you get to this step, I can help you isolate some of this, or at least show some examples.

Also note the comment within the client strings file:

    // Editor strings, none are used since there's some kind of literal string dependency - that should be fixed

    static readonly scriptTab = 'Script';
    static readonly htmlTab = 'Template';
    static readonly cssTab = 'Style';
    static readonly librariesTab = 'Libraries';
    static readonly scriptTabTag = Strings.scriptTab.toLowerCase();
    static readonly htmlTabTag = 'template';
    static readonly cssTabTag = Strings.cssTab.toLowerCase();
    static readonly librariesTabTag = Strings.librariesTab.toLowerCase();

We can remove the dependency, but for now start with assuming that the tab names in particular should not be changed.

And beyond that, commit early and often :-)

Have fun, and do reach out if you have questions.  For me personally, I found that venturing into web tech was really really fun, albeit very confusing at first.  I hope you'll find the experience equally gratifying -- and if this will make Script Lab more applicable and more appealing to a section of less English-comfortable users, that's wonderful!

Thanks!
- Michael

**Comment by maninweb** [comment]
Hello Michael,

great, that you like the idea to having localized versions of Script Lab in the future. I think, that this will be surely helpful for many users. Yes, I can treat the localization as an expriment and I am very excited to learn and contribute. I have some experience in HTML, CSS and PHP, but techniques such as modern JS, Angular and many of tools such as npm or node.js are new for me.

So, many thanks for your explanations and guidance, I really appreciate this. I am going to start first to get it run locally and will then of course come back here.

Thanks,
Mourad

**Comment by maninweb** [comment]
Hello Michael,

I got the local installation running by following the steps from contribute.md :-)

I would like to share my experience while following the steps from contribute.md. As I had already installed Visual Studio Code and Node JS, I could skip these steps. Installing Yarn worked, but when trying to run it, I firstly got nothing. Until I noticed that Yarn adds a path to the system environment path and I had to log off and log in to my system (Windows 7 in this case). Then all worked fine.

When running the server on localhost, I got the expected certificate error. The animated gif in the contribute.md is a little bit fast (speed) and therefore it's somehow difficult to follow the exact steps shown in the animated gif. One step shows a view from the IE import wizard with "Current user" and "Local machine". I did not get this option in my wizard (IE 11). I went further and added the localhost certificate to my Trusted Root Certificate Authorities. Worked. The IE now does not present the warning page when I call the URL, but a one time message box that the certificate has been issued by a company I do not trust. If I click on Yes to continue and ignore the warning, then the message does not appear again in the browser during the active session.

![image](https://cloud.githubusercontent.com/assets/13045898/25628243/d1f327a4-2f65-11e7-9d81-12bb8b0b1a63.png)

I tried several things to get rid of this message:
- redo the steps, including to run the IE explicitely as admin,
- export the certificate from the IE, open the management console and import it there to the local computer in Trusted Root Certificate Authorities,
- changing the security levels in the IE settings for the trusted sites to low (was on middle by default).

Nothing worked, the IE still shows me the message box when I call my localhost for the first time. Interestingly, in Chrome (my default browser) all works fine and I have no problem. 

Copying then the manifest for local use in a shared trusted folder for installing and then letting run the local Add-In in Excel 2016 worked fine too. However, as soon as I open the Code task pane in Script Lab [dev] in Excel, I get the message box from the IE again. If I do not respond quickly (by clicking on Yes), another one pops up and so on. Once I click on Yes (for all open popups), then the message does not appear again in the session.

Enabling auth for local testing works for me well too. I also tested some of the manual testing scenarios mentionned in contribute.md. Creating new snippets or snippets from a sample works for me, importing from and sharing to Github too. Intellisense, Syntax Highlighting, Errors, Scrolling in the Console all work for me. However, when I run a script in the Run Task Pane, I get two compiler warnings:

```
[WDS] Warnings while compiling.
./~/@angular/core/@angular/core.es5.js
5889:15-36 Critical dependency: the request of a dependency is an expression
./~/@angular/core/@angular/core.es5.js
5905:15-102 Critical dependency: the request of a dependency is an expression
```

Best,
Mourad

**Comment by Zlatkovsky** [comment]
Hi Mourad,

Glad you were able to get the initial step working!  For comments re. the experience:  might I suggest you put in a pull request to the contributing doc?

I am not sure offhand about the localhost issue, I'd seen it on another computer too.  Sounds like it's not blocking, though if I see it again, I'll try to debug.  It doesn't happen on my two computers that I tried this on.

Re. the "run" warnings:  yes, good call, I think they always appear in debug mode, and are 100% innocent.  If you do a pull request to the contributing doc, could you add a note that seeing those in "run" mode while on localhost is completely normal?

Thanks!

**Comment by maninweb** [comment]
Hi Michael,

I created a pull request for the contribute.md file from my fork of Script Lab today.

While testing, I noticed the following: when I run Script Lab from VS Code, I use npm start and then enter to the debug mode with F5. When I am finished, I use Shift+F5 to stop debugging. Each time then, I get the message "[nodemon] app crashed, waiting for file changes before starting..." in my terminal. Is this normal? If I try to restart, I get an error in VS Code "Cannot connect to runtime via 'legacy' protocol; consider using 'inspector' protocol (timeout after 10000 ms)." I have to restart VS Code for getting F5 to work again.

Best,
Mourad

**Comment by Zlatkovsky** [comment]
Thanks for the pull request!  I merged it it now just now.

For stopping debugging, added the following to the CONTRIBUTING.md(https://github.com/OfficeDev/script-lab/blob/master/CONTRIBUTING.md) doc:

> Note: To stop debugging, first press the "stop" button on the debugger (shift + F5). You will see an error on the server terminal window ([nodemon] app crashed - waiting for file changes before starting...); that's ok, just close that terminal instance (trash icon on top-right of terminal). If you want to restart debugging later, just re-press F5 again.

**Comment by maninweb** [comment]
Hello Michael,

the last days, I spent time to get more familiar with Github and VS Code. After reading a lot of documentation, testing things accross my accounts on GitHub, I ended by creating a [branch on my Script Lab fork](https://github.com/maninweb/script-lab/tree/translation-german-formal) where I am committing all my translations for Script Lab. I hope this is the correct way.

I commited the server/core/strings.ts file there, the simpliest one to translate. I am currently on the client/app/helpers/strings.ts file which also has a commit but is incomplete for now.

I have a question regarding the client strings file: is there a way to invoke error messages by code (may be as a snippet within Script Lab?) so that I can see how it looks in German? In German, some sentences may become much more longer than in English and I would like to see how the error appears in Script Lab. For example, _static readonly reloadPrompt_ from _client/app/helpers/strings.ts_ is longer than in English.

The enabling auth for local testing works fine for me. However, if I change the files env.config.js and server.ts, VS Code marks them as changed in the Source Control. I thought, may be it's possible to let VS Code ignore them and never commit the changes to my fork. I tried to change VS Code settings, add them to .gitignore, but this does not work, meaning I could not get these files never synced. Do you perhaps have a quick tipp, how I can achieve this? Thanks! If it is possible, what do you think about to add the tipp to the contribute.md?

Best,
Mourad

**Comment by Zlatkovsky** [comment]
The fork approach is correct.

For seeing the code in action:  I'm confused, shouldn't you be able to just run the add-in and see the changes right there?  I suppose it may not always be 100% obvious what triggers the text, but if you search for whether the thing is used, you'll generally be able to tell.  For example, searching for `reloadPrompt`, I see it in the following:

    @Effect()
    import$: Observable<Action> = this.actions$
        .ofType(Snippet.SnippetActionTypes.IMPORT)
        .map((action: Snippet.ImportAction) => ({ data: action.payload, mode: action.mode }))
        .mergeMap(({ data, mode }) => this._importRawFromSource(data, mode), ({ mode }, snippet) => ({ mode, snippet }))
        .filter(({ snippet }) => !(snippet == null))
        .mergeMap(({ snippet, mode }) => this._massageSnippet(snippet, mode))
        .catch((exception: Error) => {
            const message = (exception instanceof PlaygroundError) ? exception.message : Strings.snippetImportErrorBody;
            this._uiEffects.alert(
                message + '\n\n' + Strings.reloadPrompt,
                Strings.snippetImportErrorTitle,
                Strings.okButtonLabel)
                .then(() => window.location.reload());
            return Observable.from([]);
        });

This means that to see it in action, simply pass in an invalid URL, and you should see the alert show up. 
 Does that make sense?

For the local auth testing -- you should only need to do this optionally when testing out auth scenarios (I added a clarifying statement to it yesterday).  So generally, I would say don't worry about it,and just revert those changes.  You can apply them temporarily (and then revert) when you want to test some auth-related strings.

- Michael

**Comment by maninweb** [comment]
Hello Michael,

thank you, yes it makes of course sense and I could reproduce it by passing an invalid URL to the Gist. The Add-in run perfectly on localhost, no problem there.

Sorry, that I confused you.  I think, I did not explain exactly yesterday, why I would like to throw errors by code and may be the example for `reloadPrompt` was not a good one. I had in mind and a feeling, that too long sentences in German may perhaps break the UI. That's why I thought, seeing the results in the UI for _each_ error would be helpful. Yes, I can invoke some errors by the UI, but not all, e.g. `intellisenseClearError`. But after the search you recommended, I have seen that `intellisenseClearError` calls `UI.ReportErrorAction` like `githubLoginFailed`. This one was easy to trigger by just not login to GitHub. So, I've seen the dialogue and could do tests with this one. This is suffisent for me and I think, no more necessary to see the dialogue for the others.

Today, I tested two errors: `reloadPrompt` and just a dummy string for `githubLoginFailed` and noticed the following. When a sentence is too long, then I can see that words are broken at the end of a line, if e.g. the error message has two lines. For a large Task Pane, this is character position 54. For example for `reloadPrompt` with long sentences I see this:

![image](https://cloud.githubusercontent.com/assets/13045898/25772123/a5cffdc2-3263-11e7-86b9-a946aa044254.png)

I tried `githubLoginFailed` too. Putting in a long dummy sentence leads to the word broken at character 54 too in an enlarged task pane. 

If I resize the task pane by decreasing the width, then at a certain point, the error message gets smaller in width too, as it is responsive. This is fine. But if I resize it to small width, this leads to that also words in the English original error messages would be broken. I think, nicier would be that it would break at the white space before (also for other languages with longer texts then). What do you think? May be this is something in the CSS to be changed? Should I create an issue for this, separately from this thread?

Best,
Mourad

**Comment by Zlatkovsky** [comment]
For the dialog breaking words:  yes, I noticed it before as well, it (the dialog) was hastily created, and has some issues.  Please file a separate issue and I'll see if there's a simple fix we can do.

For the rest:  let me know when you're ready for me to try to integrate your changes (though i probably won't get to it till the end of the week -- there are some changes I'd like to get in before //build, and after that I want to keep things stable for a few days during the conference).

**Comment by maninweb** [comment]
Hello Michael,

I have now translated all files and uploaded them to my fork here: https://github.com/maninweb/script-lab/tree/translation-german-formal.

I added the new document _translations.md_ where I currently list which files I modified and which not and where I did some notes about the translation process. I thought to later extend this document with more details, so that e.g. if someone else would like to translate Script Lab, a how to would be available.

Best,
Mourad

**Comment by Zlatkovsky** [comment]
@maninweb , looks great!

@saunders77 , @jakobpn , I wonder if you want to team up with me (read: do the work :-)) of integrating in the translation.  We have some documentation at https://dev.office.com/docs/add-ins/develop/localization that we can use to dogfood that doc, and I'm happy to help provide ideas/review; I just don't have the bandwidth to do it myself (or at least, any bandwidth here would be taking away time from other features).

**Comment by Zlatkovsky** [comment]
@maninweb , just to let you know:  we've not forgotten about you!  we've just all been catching up on our day-to-day work (Script Lab was a side project for all of us...)

Jakob & Michael Saunders & I met this afternoon to establish a plan.  It sounds like the first step is for me to do some of the more technical work of hooking up the translation (so as a proof-of-concept, just offer English and English -- but then be able to have Jakob (or yourself, if you're willing) help out with taking your translations and moving them over to the system.

I'll try to do this in the coming week or two...

**Comment by maninweb** [comment]
Hi Michael,

thanks, these are great news :-) Yes, if I would be glad to help and support you.

Best,
Mourad

**Comment by Zlatkovsky** [comment]
@maninweb , just FYI that we still haven't forgotten about this.  I've been going through and adding a "fake" localization, and moving a bunch of text into the Strings file, so that we can then support actual languages too.

See the screenshot below, for example.

![image](https://user-images.githubusercontent.com/2230453/27778490-255c6574-5f70-11e7-9452-2fd655226d6d.png)

I'll let you know as soon as we've integrated your changes.

**Comment by maninweb** [comment]
Good morning/evening Michael, great :-)

**Comment by Zlatkovsky** [comment]
@maninweb , @jakobpn :  in what took _way more infrastructure_ than I originally anticipated, I have finally finished the change to support localization, tested it with a set of fake strings, and merged it in to the master branch of the OfficeDev/script-lab repo.

@maninweb , I did not yet incorporate your German translations, but it should now be very easy.  All that it would take is to add a `german.ts` file, copying from the `english.ts` file, in `src\client\app\strings` and `src\server\strings` (two separate string files for server vs. client).  Given the amount of changes that the repo has undergone in that infrastructure change, it will probably be easiest to just copy-paste the needed strings out of your diff (which is viewable via https://github.com/OfficeDev/script-lab/compare/master...maninweb:translation-german-formal).  I believe @jakobpn was willing to help, since you did the initial up-front work.  The strings are now all in one(two) places instead of in the various html pages and Angular markup, so it should be quite straightforward.  And once the `german.ts` files are created, it takes literally two lines of code in the `index.ts` for each of these folders, specifying that German is now supported in addition to English.

By the way, I will say that even though I didn't use your change directly (since I wasn't incorporating the German yet, just building the infrastructure to support localization), it was supremely useful to have your diff to look at and see all the many places where we had strings scattered in the product code, as well as looking at your `translation.md` doc.  It should also make Jakob's job quite easy.  (BTW, there are a number of new strings that have materialized in the meantime, so those will need translating too)

For incorporating the localization in the manifest, @jakobpn or @saunders77 , could you make the necessary changes (incorproating Mourad's strings as localized resources)?  It's a process that is documented under https://dev.office.com/docs/add-ins/develop/localization.  I will leave it in your capable hands :-)

Thanks, all!

~ Michael

**Comment by maninweb** [comment]
Hi all,

thank you, Michael, for all this work - it's now much more easier to translate. And, I am glad, that the diff and the overview with the files list was helpful :-)

I have updated my master branch with the latest changes of your master branch and created a new branch with all translations in it. Please see here: https://github.com/maninweb/script-lab/tree/german

In this new branch "german", I added the new german.ts files, changed the index.ts files and added the manifest files to (overwritten versions of the english files). And I have also created a new [Translation.md](https://github.com/maninweb/script-lab/blob/german/Translation.md) file, including some notes too.

I did not test the translations yet (especially the new ones). I think, I will do this tomorrow.

Best,
Mourad

**Comment by Zlatkovsky** [comment]
Looks great, and thanks for your enthusiasm!  Please feel free to submit a pull request, initially with just the  TypeScript (non-manifest) files.  I'll leave the manifest portion of it to @jakobpn or @saunders77 to figure out.

A few comments:

- Before you submit, please undo the changes to the git-ignore file.  (In fact, I wonder if you'll need to even for running it, as there may have been a few changes to `env.config.js` since 2 months ago).
- If you do decide to translate "script-lab-tutorial.xlsx" and "src/client/assets/images/tutorial.png", we can make it work.  But I think the return-on-investment here might be much smaller...
- Feel free to remove the `/** NEEDS STRING REVIEW */` comments.  They are there in the English one so that @saunders77 can make the English be consistent with the other English strings (and since he's much better than me at being concise).  But if the strings sound good to your German ear, then there is nothing left to string-review.

Looking forward to taking in the pull request!  We are actively working on a number of other Script Lab features, including the ability to easily switch between testing environments, to give us better agility and confidence in shipping to production.  So I expect that your translation work should be making it into production in a couple weeks, once you submit the pull request.

Thanks again!

**Comment by Zlatkovsky** [comment]
Hi Mourad,

I just merged in the *.ts file pull request, https://github.com/OfficeDev/script-lab/pull/471.  Thanks!

For the missing string, thanks for pointing it out.  I just committed a fix:  https://github.com/OfficeDev/script-lab/commit/fc688c95324c61db5556cb95474440035c618db2.  I'm also filing a running issue https://github.com/OfficeDev/script-lab/issues/472 to track incremental translations of strings (like this "Info" one, as we need them).

For no hover titles:  yes, please file a bug.

For multiple instances mirroring the behaviors:  this only happens on a dev box (running on localhost) due to Browser Sync, which can be a mixed blessing, and not something we've intentionally made work or not.  It just happens, and only for some elements.  I wouldn't worry about it.  This will not happen in production, only when running locally.

**Comment by maninweb** [comment]
Hi Michael,

thanks :-)

I will update my fork, translate the missing strings, pull it and re-test the hover bug in the latest version.
If it's still present, I will post an issue.

Best,
Mourad

**Comment by maninweb** [comment]
@Zlatkovsky - Hi Michael, I have started to write a tutorial how to translate Script Lab. Would you please have a look https://github.com/maninweb/script-lab/blob/documents/TRANSLATING.md? What do you think? May it be useful?
Best,
Mourad

**Comment by Zlatkovsky** [comment]
@maninweb , I think it looks great!  Could you go ahead and submit it as a pull request?  I had some comments, but it will be easiest to write them out when I can just comment on the source file of the pull request.  Thanks!

**Comment by Zlatkovsky** [comment]
Looks like this issue was accidentally left open before.  Closing it.
